linter_driver_rustc/src/ast/rustc.rs

```diff
@@ -26,7 +28,68 @@ impl<'ast, 'tcx> std::fmt::Debug for RustcContext<'ast, 'tcx> {
     }
 }
 
-impl<'ast, 'tcx> DriverContext<'ast> for RustcContext<'ast, 'tcx> {}
+fn to_leaked_rustc_lint<'ast>(
+    map: &mut FxHashMap<&'ast Lint, &'static RustcLint>,
+    lint: &'ast Lint,
+) -> &'static RustcLint {
+    map.entry(lint).or_insert_with(|| {
+        Box::leak(Box::new(RustcLint {
+            name: lint.name,
+            default_level: match lint.default_level {
+                Level::Allow => RustcLevel::Allow,
+                Level::Warn => RustcLevel::Warn,
+                Level::Deny => RustcLevel::Deny,
+                Level::Forbid => RustcLevel::Forbid,
+                _ => unreachable!("added variant to lint::Level"),
+            },
+            desc: lint.explaination,
+            edition_lint_opts: None,
+            report_in_external_macro: match lint.report_in_macro {
+                MacroReport::No | MacroReport::Local => false,
+                MacroReport::All => true,
+                _ => unreachable!("added variant to lint::MacroReport"),
+            },
+            future_incompatible: None,
+            is_plugin: true,
+            feature_gate: None,
+            crate_level_only: false,
+        }))
+    })
+}
+
+impl<'ast, 'tcx> DriverContext<'ast> for RustcContext<'ast, 'tcx> {
+    fn emit_lint(&self, s: &str, lint: &'ast Lint) {
```

Comment:
Small note about the function signature: we should think about users create diagnostics. Clippy has several util functions like `span_lint`, `span_lint_and_note` etc. This forces the user to always create the suggestion before the emission, even it the lint is allowed. This can also cause some ICEs, which is why Clippy wants to move away from this model. (There is an issue for this, even if there has been no progress AFAIK).

I like the diagnostic build approach of rustc. For now, this is a good starting point. I'll be busy over the next two weeks, but this is one thing I want to discuss or at least create an issue for in the design repo. :upside_down_face:

Reply:
I'm not sure I understand what could be the other way of doing it. Or maybe I'm misunderstanding what you mean by suggestion and emission? An issue sounds like a good idea anyway better place to continue the discussion!